src/detect-dataset.c

```diff
@@ -59,6 +60,49 @@ void DetectDatasetRegister (void)
     sigmatch_table[DETECT_DATASET].Free  = DetectDatasetFree;
 }
 
+/*
+    1 match
+    0 no match
+    -1 can't match
+ */
+static int DetectDatajsonBufferMatch(DetectEngineThreadCtx *det_ctx, const DetectDatasetData *sd,
+        const uint8_t *data, const uint32_t data_len)
+{
+    if (data == NULL || data_len == 0)
+        return 0;
+
+    switch (sd->cmd) {
+        case DETECT_DATASET_CMD_ISSET: {
+            // PrintRawDataFp(stdout, data, data_len);
+            DataJsonResultType r = DatajsonLookup(sd->set, data, data_len);
+            SCLogDebug("r found: %d, len: %zu", r.found, r.json.len);
+            if (!r.found)
+                return 0;
+            if (r.json.len > 0) {
```

Comment:
this looks like a thread safety violation. `DatajsonLookup` unlocks the dataset hash entry, but the code here still references data in the hash

Reply:
Fixed.